go/types/simplify.go

```diff
@@ -191,68 +167,51 @@ func simplifyMaps(ts typeset, merge bool) *Type {
 		valTypes.Add(desc.ElemTypes[1])
 	}
 	return MakeMapType(
-		makeSimplifiedTypeImpl(keyTypes, merge),
-		makeSimplifiedTypeImpl(valTypes, merge))
+		makeSimplifiedTypeImpl(keyTypes),
+		makeSimplifiedTypeImpl(valTypes))
 }
 
-func simplifyStructs(expectedName string, ts typeset, merge bool) *Type {
-	if merge {
-		return mergeStructs(expectedName, ts)
+func simplifyStructs(expectedName string, ts typeset) *Type {
+	// We gather all the fields/types into commonFields. If the number of
+	// times a field name is present is less that then number of types we
+	// are simplifying then the field must be optional.
+	// If we see an optional field we do not increment the count for it and
+	// it will be treated as optional in the end.
+	type fieldTypeInfo struct {
+		count   int
+		typeset typeset
 	}
-	commonFields := map[string]typeset{}
+	commonFields := map[string]fieldTypeInfo{}
```

Comment:
`commonFields` is no longer a good name for this since it's no longer only the fields that the structs have in common. It's more like `fieldCounts`.